[lldb] Support parsing the Wasm symbol table

JDevlieghere opened the pull request:
This PR adds support for parsing the WebAssembly symbol table. The symbol table is encoded in the "names" section and contains names and indexes into other sections. For now we only support parsing function (code) symbols. The result is that you can set breakpoints by symbol name, while previously breakpoints by name required debug info (DWARF).

This is also necessary for Swift, which checks for the presence of `swift_release` as a heuristic to determine if there's a static Swift stdlib.

adrian-prantl commented on lldb/source/Plugins/ObjectFile/wasm/ObjectFileWasm.cpp:
```diff
@@ -248,12 +248,93 @@ bool ObjectFileWasm::ParseHeader() {
+    // function. We can't do that here because that would break symbolic
+    // breakpoints, as that address is never executed.
+    functions.emplace_back(code_section_sp, offset, function_size);
+    offset += function_size;
```
This should probably use an llvm addWithOverflow function to detect if we wrap around when reading garbage?:

adrian-prantl commented on lldb/source/Plugins/ObjectFile/wasm/ObjectFileWasm.cpp:
```diff
@@ -248,12 +248,93 @@ bool ObjectFileWasm::ParseHeader() {
+  llvm::DataExtractor data = name_section_data.GetAsLLVM();
+  llvm::DataExtractor::Cursor c(0);
+  uint32_t sym_id = 0;
+  while (c && c.tell() < data.size()) {
```
out of curiosity: isn't there a way to get a Cursor that is tied to data, that evaluates to false when it's out of space?

JDevlieghere replied:
The operator bool checks if the cursor is in an error state. I could rely on that here, but I want to be able to distinguish between an actual failure and just reading until we have no bytes left.

adrian-prantl reviewed (commented):
Looks good, I left a few comments with potential extra opportunities for error handling.

JDevlieghere committed:
[lldb] Support parsing the Wasm symbol table #153093

This PR adds support for parsing the WebAssembly symbol table. The
symbol table is encoded in the "names" section and contains names and
indexes into other sections. For now we only support parsing function
(code) symbols. The result is that you can set breakpoints by symbol
name, while previously breakpoints by name required debug info (DWARF).

This is also necessary for Swift, which checks for the presence of
swift_release as a heuristic to determine if there's a static Swift
stdlib.

JDevlieghere commented on lldb/source/Plugins/ObjectFile/wasm/ObjectFileWasm.cpp:
```diff
@@ -248,12 +248,93 @@ bool ObjectFileWasm::ParseHeader() {
+  functions.reserve(function_count);
+
+  for (uint32_t i = 0; i < function_count; ++i) {
+    const uint32_t function_size = code_section_data.GetULEB128(&offset);
```
Looking at the implementation in LLVM, it returns `0`.

DavidSpickett commented on lldb/source/Plugins/ObjectFile/wasm/ObjectFileWasm.cpp:
```diff
@@ -248,12 +249,136 @@ bool ObjectFileWasm::ParseHeader() {
+    std::optional<lldb::offset_t> next_offset =
+        llvm::checkedAddUnsigned(offset, function_size);
+    if (!next_offset)
+      return llvm::createStringError("offset overflows uint64_t");
```
offset -> function offset

JDevlieghere committed:
offset -> function offset

DavidSpickett reviewed (approved):
> The result is that you can set breakpoints by symbol name, while previously breakpoints by name required debug info (DWARF).

Assuming that this does not break things if both are present, this LGTM.
